### PR TITLE
Adjust XMLHttpRequest Content-Type handling

### PR DESCRIPTION
**Pull request by annevk**
See https://github.com/whatwg/mimesniff/pull/36.

### CODE REVIEW

**Comment by annevk** [comment]
Always replacing with uppercase UTF-8 could be considered I think. I think Firefox does not do that because it broke some sites, but those may well be gone/fixed now.

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/21031)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/306798513)

**Comment by annevk** [comment]
TODO: tests with double quotes.

**Comment by annevk** [comment]
TODO: update XMLHttpRequest to actually invoke the MIME type parser and serializer when doing this manipulation.

**Comment by annevk** [comment]
It does not seem that Chrome always replaces with UTF-8 by the way. And Firefox replaces a whole lot with UTF-8 as it seems it doesn't do double quotes correctly. This makes it a little bit unclear to me what the processing model should be here. Anyone ideas?

**Comment by annevk** [comment]
This is ready now. This is what is remaining:

- [ ] Merging of https://github.com/whatwg/mimesniff/pull/36 (the new MIME type specification)
- [ ] Usage of that MIME type specification in the XMLHttpRequest Standard

**Comment by annevk** [comment]
This could use review now.

File: xhr/send-content-type-charset.htm [hunk 1]
```diff
@@ -70,14 +75,30 @@
...
-        "text/x-pink-unicorn; charset=UTF-8; charset=UTF-8; notrelated; charset=UTF-8",
-        "If multiple charset parameters are given, all should be rewritten"
+        "text/x-pink-unicorn;charset=UTF-8",
+        "Multiple charset parameters deduplicate, bogus parameter dropped"
       )
       request(
         null,
         "text/plain;charset=UTF-8",
         "No content type set, give MIME and charset"
       )
```
Comment (domenic) [inline xhr/send-content-type-charset.htm]: Is it worth testing that the empty byte sequence gets passed through? (Note you'd have to change the check in request from `if (input)` to `if (input !== null)`.)

Comment (domenic) [inline xhr/send-content-type-charset.htm]: Might be worth testing the other cases of HTML document and XML document. Although I guess they are probably tested elsewhere.

File: xhr/send-content-type-charset.htm [hunk 2]
```diff
@@ -70,14 +75,30 @@
...
+        "charset in double quotes")
+      request(
+        "text/plain;charset=\" utf-8\"",
+        "text/plain;charset=UTF-8",
+        "charset in double quotes with space")
+      request(
+        "text/plain;charset=\"u\\t\\f-8\"",
+        "text/plain;charset=UTF-8",
+        "charset in double quotes with backslashes")
     </script>
```
Comment (domenic) [inline xhr/send-content-type-charset.htm]: Can you add tests that non-charset parameters also get preserved (and parsed/re-serialized), when a charset parameter is present?

**Review by domenic — APPROVED** [review]
LGTM but I think at least the tests with non-charset parameters alongside charset should probably be added.

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/27437)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/367065238)

**Comment by benbucksch** [comment]
re changing charset=utf-8 to charset=UTF-8:

I am trying to do autodiscover with outlook.com, and use XmlHttpRequest(), and set the Content-Type to "text/xml; charset=utf-8". Due to this bug, outlook.com server (<https://outlook.com/autodiscover/autodiscover.xml> with POST of XML) tells me:
Cannot process the message because the content type 'text/xml;charset=UTF-8' was not the expected type 'text/xml; charset=utf-8'.

Now, I agree that they are wrong and the space is optional and charsets are case insensitive. The (old) spec is clear on that and outlook.com is misbehaving. Yet, this is preventing me from contacting the service, and XmlHttpRequest doesn't let me pass the header as expected anymore.

xref https://bugzilla.mozilla.org/show_bug.cgi?id=1454325

**Pull request merged**